stability_image/image.go

```diff
@@ -0,0 +1,392 @@
+			}
+			rightAnchor := image.Point{
+				X: centeredSourceBounds.Max.X,
+				Y: centeredSourceBounds.Min.Y,
+			}
+			target = imaging.Overlay(target, reflection, leftAnchor, 1)
+			target = imaging.Overlay(target, reflection, rightAnchor, 1)
+		case DirectionLeft:
+			leftAnchor := image.Point{
+				X: sourceDim.X,
+				Y: 0,
+			}
+			target = imaging.Overlay(target, reflection, leftAnchor, 1)
+		case DirectionRight:
+			rightAnchor := image.Point{
+				X: targetDim.X - (sourceDim.X * 2),
+				Y: 0,
+			}
+			target = imaging.Overlay(target, reflection, rightAnchor, 1)
+		}
+	}
+	// If requested, blur the edges to preserve continuity
+	if blur > 0 {
+		var blurErr error
+		if target, blurErr = stackblur.Process(target, blur); blurErr != nil {
+			return nil, blurErr
+		}
+	}
+	return target, nil
+}
+
+func QuantizePng(old *[]byte, quantization int) (new *[]byte, err error) {
+	img, _, err := image.Decode(bytes.NewReader(*old))
+	if err != nil {
+		return nil, err
+	}
+	compressed := lossypng.Compress(img, lossypng.NoConversion, quantization)
+	return EncodePng(compressed, png.BestSpeed)
+}
+
+func CreateBorder(width uint64, height uint64) (border *[]byte, err error) {
+	// Create a new image
+	img := image.NewRGBA(image.Rect(0, 0, int(width), int(height)))
+	// Draw the border
+	Fill(img, &color.RGBA{A: 255})
+	// Encode the image
+	buf := new(bytes.Buffer)
+	err = png.Encode(buf, img)
+	if err != nil {
+		return nil, err
+	}
+	borderBytes := buf.Bytes()
+	return &borderBytes, nil
+}
+
+func CreateGradient(
+	outpaintGradient colorgrad.Gradient,
+	gradientDim image.Rectangle,
+	gradientDirection Direction,
+) (i *image.RGBA) {
+	i = image.NewRGBA(gradientDim)
+	var gradientColorFn func(x, y int) colorful.Color
+	switch gradientDirection {
+	case DirectionUp:
+		gradientColorFn = func(x, y int) colorful.Color {
+			scaled := float64(y) / float64(gradientDim.Max.Y)
+			return outpaintGradient.At(scaled)
+		}
+	case DirectionDown:
+		gradientColorFn = func(x, y int) colorful.Color {
+			scaled := float64(y) / float64(gradientDim.Max.Y)
+			return outpaintGradient.At(1 - scaled)
+		}
+	case DirectionLeft:
+		gradientColorFn = func(x, y int) colorful.Color {
+			scaled := float64(x) / float64(gradientDim.Max.X)
+			return outpaintGradient.At(scaled)
+		}
+	case DirectionRight:
+		gradientColorFn = func(x, y int) colorful.Color {
+			scaled := float64(x) / float64(gradientDim.Max.X)
+			return outpaintGradient.At(1 - scaled)
+		}
+	case DirectionCenter:
+		if gradientDim.Max.Y > gradientDim.Max.X {
+			gradientColorFn = func(x, y int) colorful.Color {
+				scaled := float64(x) / float64(gradientDim.Max.X)
+				return outpaintGradient.At(1 - scaled)
+			}
+		} else {
+			gradientColorFn = func(x, y int) colorful.Color {
+				scaled := float64(y) / float64(gradientDim.Max.Y)
+				return outpaintGradient.At(1 - scaled)
+			}
+		}
+	}
+	for x := 0; x < i.Rect.Max.X; x++ {
+		for y := 0; y < i.Rect.Max.Y; y++ {
+			i.Set(x, y, gradientColorFn(x, y))
```

Comment:
This is doing a lot of work per pixel that could be done per row/column.

Reply:
Agreed. But if you look at the Imaging source code, it does it there too. 🤔